build-tools/packages/build-cli/src/library/packageExports.ts

```diff
@@ -0,0 +1,163 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import path from "node:path";
+
+import type { Logger, PackageJson } from "@fluidframework/build-tools";
+
+/**
+ * Properties for an "exports" leaf entry block in package.json.
+ * A block is the set of conditions and final results as in:
+ * ```json
+ * {
+ *    "types": "./index.d.ts",
+ *    "default": "./index.js"
+ * }
+ * ```
+ */
+export interface ExportData {
+	/**
+	 * Location of file relative to package
+	 */
+	relPath: string;
+	/**
+	 * Conditions required for export access; in hierarchy order
+	 */
+	conditions: string[];
+	/**
+	 * Export is only .d.ts file
+	 */
+	isTypeOnly: boolean;
+}
+/**
+ * Minimal set of properties required from an "exports" entry to generate
+ * Node10 compatible redirection files.
+ */
+export type Node10CompatExportData = Pick<ExportData, "relPath" | "isTypeOnly">;
+
+/**
+ * Only the value types of exports that are records.
+ */
+type ExportsRecordValue = Exclude<Extract<PackageJson["exports"], object>, unknown[]>;
+
+function findTypesPathMatching<TOutKey>(
+	mapQueryPathToOutKey: Map<string | RegExp, TOutKey | undefined>,
+	exports: ExportsRecordValue,
+	conditions: string[],
+): (ExportData & { outKey: TOutKey | undefined }) | undefined {
+	for (const [entry, value] of Object.entries(exports)) {
+		if (typeof value === "string") {
+			if (entry === "types") {
+				for (const [key, outKey] of mapQueryPathToOutKey.entries()) {
+					// eslint-disable-next-line max-depth
+					if (
+						typeof key === "string"
+							? path.resolve(value) === path.resolve(key)
+							: key.test(value)
+					) {
+						const isTypeOnly = !(
+							"default" in exports ||
+							"import" in exports ||
+							"require" in exports
+						);
+						return { outKey, relPath: value, conditions, isTypeOnly };
+					}
+				}
+			}
+		} else if (value !== null) {
+			if (Array.isArray(value)) {
+				continue;
+			}
+			const deepFind = findTypesPathMatching(mapQueryPathToOutKey, value, [
+				...conditions,
+				entry,
+			]);
+			if (deepFind !== undefined) {
+				return deepFind;
+			}
+		}
+	}
+
+	return undefined;
+}
+
+/**
+ * Read package "exports" to determine which of given file paths are present.
+ *
+ * @param packageJson - json content of package.json
+ * @param mapQueryPathToOutKey - keys of map represent paths to match. When matched
+ * value, if defined, is used to set entry key in output mapKeyToOutput.
+ * @param node10TypeCompat - when true, populates output mapNode10CompatExportPathToData.
+ * @param logger - optional Logger
+ * @returns object with mapKeyToOutput, map of ApiTags to output paths, and
+ * mapNode10CompatExportPathToData, map of compat file path to Node16 path.
+ */
+export function queryOutputMapsFromPackageExports<TOutKey>(
+	packageJson: PackageJson,
+	mapQueryPathToOutKey: Map<string | RegExp, TOutKey | undefined>,
```

Comment:
Can this be a `ReadonlyMap` since we don't appear to be modifying it here?

Reply:
Good call